python/dgl/graphbolt/impl/ondisk_dataset.py

```diff
@@ -38,6 +41,215 @@
+        Whether to include the original edge id in the FusedCSCSamplingGraph.
+
+    Returns
+    -------
+    sampling_graph : FusedCSCSamplingGraph
+        The FusedCSCSamplingGraph constructed from the raw data.
+    """
+    is_homogeneous = "type" not in graph_data["nodes"][0]
+
+    if is_homogeneous:
+        # Homogeneous graph.
+        edge_fmt = graph_data["edges"][0]["format"]
+        edge_path = graph_data["edges"][0]["path"]
+        src, dst = read_edges(dataset_dir, edge_fmt, edge_path)
+        num_nodes = graph_data["nodes"][0]["num"]
+        num_edges = len(src)
+        coo_tensor = torch.tensor([src, dst])
+        sparse_matrix = dglsp.spmatrix(coo_tensor)
+        indptr, indices, value_indices = sparse_matrix.csc()
+        node_type_offset = None
+        type_per_edge = None
+        node_type_to_id = None
+        edge_type_to_id = None
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            edge_attributes[ORIGINAL_EDGE_ID] = value_indices
+    else:
+        # Heterogeneous graph.
+        # Construct node_type_offset and node_type_to_id.
+        node_type_offset = [0]
+        node_type_to_id = {}
+        for ntype_id, node_info in enumerate(graph_data["nodes"]):
+            node_type_to_id[node_info["type"]] = ntype_id
+            node_type_offset.append(node_type_offset[-1] + node_info["num"])
+        # Construct edge_type_offset, edge_type_to_id and coo_tensor.
+        edge_type_offset = [0]
+        edge_type_to_id = {}
+        coo_src_list = []
+        coo_dst_list = []
+        coo_etype_list = []
+        for etype_id, edge_info in enumerate(graph_data["edges"]):
+            edge_type_to_id[edge_info["type"]] = etype_id
+            edge_fmt = edge_info["format"]
+            edge_path = edge_info["path"]
+            src, dst = read_edges(dataset_dir, edge_fmt, edge_path)
+            edge_type_offset.append(edge_type_offset[-1] + len(src))
+            src_type, _, dst_type = etype_str_to_tuple(edge_info["type"])
+            src += node_type_offset[node_type_to_id[src_type]]
+            dst += node_type_offset[node_type_to_id[dst_type]]
+            coo_src_list.append(torch.tensor(src))
+            coo_dst_list.append(torch.tensor(dst))
+            coo_etype_list.append(torch.full((len(src),), etype_id))
+
+        coo_src = torch.cat(coo_src_list)
+        coo_dst = torch.cat(coo_dst_list)
+        coo_etype = torch.cat(coo_etype_list)
+
+        sparse_matrix = dglsp.spmatrix(
+            indices=torch.stack((coo_src, coo_dst), dim=0)
+        )
+        indptr, indices, value_indices = sparse_matrix.csc()
+        node_type_offset = torch.tensor(node_type_offset)
+        type_per_edge = torch.index_select(
+            coo_etype, dim=0, index=value_indices
+        )
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            edge_attributes[ORIGINAL_EDGE_ID] = value_indices
```

Comment:
original EIDs are supposed to be per-type IDs. does `value_indices` are per-type or global EIDs?

Reply:
Yes it is, and dedicated tests will be added thanks to your remind